node: Add LocalNodeStore for coordinating updates to local node state

joamaki opened the pull request:
First commit adds pkg/stream that contains the interface definition for `Observe` and few utilities. I'm including few more than used already, with the idea that observer to the LocalNodeStore would use for example the `Map` and `Distinct` combinators to only observe changes relevant to it, e.g. to only be informed if EncryptionKey changes, one would do:
```go
Distinct(
  Map(localNodeStore, func(n types.Node) uint8 { return n.EncryptionKey })
).Observe(ctx, onEncryptionKeyChanged, onCompletion)
```

Second commit adds LocalNodeStore that manages the `types.Node` for the local node and provides an API to observe (as defined by pkg/stream) changes to it and to atomically update it. To introduce this large change incrementally only the getters and setters in `pkg/node` are changed and nothing yet observes the changes. 

Follow-ups will start converting uses of getters/setters over to LocalNodeStore. The fields remaining in `node.addresses`, which are not included in `types.Node`, will eventually be provided via similar observable abstraction by the DeviceManager.

LocalNodeStore makes the assumption that `types.Node` is cheap to copy & not access too frequently and that we prefer copying over passing a pointer to `types.Node` that could be inadvertently mutated by an observer.

joamaki committed:
stream: Initial implementation of the Observable and utilities

This is the initial implementation for the Observable interface
and the utility functions around it.

This builds upon my experience of implementing
github.com/joamaki/goreactive, but changes from the blocking Observe()
into an asynchronous model in order to ensure that we can start
observing streams during module construction, without having to deal
with races.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
node: Add LocalNodeStore and convert getters/setters to it

This adds LocalNodeStore, a canonical owner of 'types.Node' for the local
node. The supported operations on the store are Observe, Update and Get.

With this construction we can simplify the persisting of the local state
and the code that modify the local node state. Typical example is ipsecKeyIdentity
updated from pkg/ipsec, where not only we need to update the global variable,
but also inform both nodediscovery (to push change to k8s/kvstore) and datapath.
Now instead we would call 'LocalNodeStore.Update' instead and can drop the IPsec
dependency to nodediscovery and NodeHandler.

Since this is a large change to push through, this starts by simply converting
the existing getters and setters over to using LocalNodeStore, without yet
observing it.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

squeed reviewed (approved):
nice work!

jibi commented on pkg/k8s/init.go:
```diff
@@ -183,9 +183,6 @@ func WaitForNodeInformation(ctx context.Context, nodeGetter nodeGetter) error {
 		node.SetK8sExternalIPv6(n.GetExternalIP(true))
 
-		// K8s Node IP is used by BPF NodePort devices auto-detection
-		node.SetK8sNodeIP(k8sNodeIP)
```
Missing `node.SetIPv4()`?

Also might be worth documenting the switch to the double setter `SetIPv{4,6}()/SetK8sExternalIPv{4,6}()` -> single getter `GetK8sNodeIP()` in a a separate commit

joamaki replied:
`SetIPv4` is on line 174. A follow up that I'm working on this week will remove these setters and instead copy the whole `types.Node` from k8s into LocalNodeStore in an initializer, so not sure it's worth the effort to bring attention to this. EDIT: actually, better play it safe. I'll move it into its own commit and explain the rational.

jibi commented on pkg/node/host_endpoint.go:
```diff
@@ -3,32 +3,26 @@
 }
 
 // SetLabels sets the labels of this node.
 func SetLabels(l map[string]string) {
```
Can these 2 functions be moved to `node.go`?

joamaki replied:
I would like to not move or rename things yet as a follow-up will remove the setters and likely mostly remove uses of the getters.

jibi commented on pkg/node/types/node.go:
```diff
@@ -332,6 +332,52 @@ func (n *Node) GetCiliumInternalIP(ipv6 bool) net.IP {
+	n.IPAddresses = newAddresses
+}
+
+func (n *Node) SetAddress(typ addressing.AddressType, newIP net.IP) {
```
Could we add setters for the rest of the address types: https://github.com/cilium/cilium/blob/a5f59b990015f81b5f5ad92f6214d9a6000abc08/pkg/node/addressing/addresstype.go#L12-L17 and unexport this?

joamaki replied:
Sure!

joamaki replied:
Hostname, InternalDNS and ExternalDNS are not IP addresses, but FQDNs, so won't add those.

jibi commented on pkg/node/address.go:
```diff
@@ -93,25 +80,29 @@ func InitDefaultPrefix(device string) {
 	defer addrsMu.Unlock()
```
Do we still need this lock here?

joamaki replied:
No we don't. I'll remove it from this PR. In a follow-up this thing will move into the "LocalNodeInitializer" which should further clean this up and stop from seeing a partial view of these fields.

jibi commented on pkg/node/address.go:
```diff
@@ -33,39 +35,20 @@ var (
 )
 
 type addresses struct {
 	ipv4Loopback      net.IP
```
Should this be moved as well to use the `LocalNodeStore`?

joamaki replied:
There is no loopback address type, so not representable in types.Node. This is also only settable via `option.Config.LoopbackIPv4` so might make sense to make this solely a config option. I'll address this later.

jibi commented on pkg/node/address.go:
```diff
@@ -297,44 +288,41 @@ func SetIPv4(ip net.IP) {
 
 // SetK8sExternalIPv4 sets the external IPv4 node address. It must be a public IP that is routable
 // on the network as well as the internet.
 func SetK8sExternalIPv4(ip net.IP) {
```
nit: rename `SetK8sExternalIPv4()` to `SetNodeExternalIPv4()` (or viceversa) for consistency?

joamaki replied:
I wasn't renaming these on purpose even though I was quite annoyed by the inconsistencies as a follow-up PR would remove these completely in favor of a LocalNodeInitializer.

jibi reviewed (commented):
Looking good and love how the stream package will make it concise to interact with the local node state! I did a first round focusing on the second commit, mostly nits and questions (will do a second round on Monday)

joamaki committed:
k8s: Remove unnecessary SetK8sNodeIP

The K8s node IP was derived from the node addresses and is now unneeded
as SetIPv4/SetIPv6 set the same address.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

aanm commented on pkg/stream/sources.go:
```diff
@@ -0,0 +1,219 @@
+// FromSlice converts a slice into an Observable.
+func FromSlice[T any](items []T) Observable[T] {
+	// Emit items in chunks to reduce overhead of mutex in ctx.Err().
+	const chunkSize = 64
```
This seems a nice random number :slightly_smiling_face:

joamaki replied:
Yeah quite random. Also premature optimization as I'm expecting `FromSlice` to just be used in tests for most parts and it's not performance critical...

aanm commented on pkg/stream/sources.go:
```diff
@@ -0,0 +1,219 @@
+			sub.complete()
+		}
+		subs = nil
+		mu.Unlock()
```
why didn't you defer here as well?

joamaki replied:
this is potentially called a lot and "defer" is 16x more expensive: https://medium.com/i0exception/runtime-overhead-of-using-defer-in-go-7140d5c40e32. 

EDIT: doh, should've looked my code closer.. this was the complete() and not next() so that reasoning makes no sense. I'll just add the defer. :D

gandro replied:
FWIW, that blog post is quite old. Go 1.14 significantly improved `defer` performance in many cases: https://rakyll.org/inlined-defers/

joamaki replied:
Ah doh, I haven't kept up with the times :)
